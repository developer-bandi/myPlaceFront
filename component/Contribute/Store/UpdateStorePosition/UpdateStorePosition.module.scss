.mainBlock {
  width: 100%;
}

.title {
  text-align: center;
  margin: 50px 0 50px 0;
}

.subBlock {
  margin: auto;
  width: 1000px;
  height: 1150px;
  margin-bottom: 20px;
  background-color: white;
  padding: 30px;
  border-bottom: 2px solid black;
  border-top: 2px solid black;
}

.subTitle {
  margin-bottom: 20px;
}
.addressBlock {
  margin: 20px 0 50px 0;
}
.addressInput {
  z-index: 1;
  width: 850px;
  height: 40px;
  background-color: #eaeaea;
  margin-top: 10px;
}

.addressButton {
  width: 70px;
  height: 40px;
  background-color: #00aaff;
  border-radius: 5px;
  color: white;
  margin-left: 20px;
}

.submitButtonBlock {
  width: 1000px;
  margin: auto;
  display: flex;
  justify-content: end;
  margin-bottom: 100px;
}
