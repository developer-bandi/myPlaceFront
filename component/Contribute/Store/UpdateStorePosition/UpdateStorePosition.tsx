import {RefObject} from "react";
import storePosition from "../../../../lib/styles/storePosition.module.scss";

interface UpdateStorePositionProps {
  mapref: RefObject<HTMLDivElement>;
  addressInputRef: RefObject<HTMLInputElement>;
  changeAddress: () => void;
  changePosition: () => void;
  isTabletOrMobile: boolean;
}

const UpdateStorePosition = ({
  mapref,
  addressInputRef,
  changeAddress,
  changePosition,
  isTabletOrMobile,
}: UpdateStorePositionProps) => {
  return (
    <div className={storePosition.mainBlock}>
      <h1 className={storePosition.title}>장소 수정</h1>
      <div
        ref={mapref}
        style={
          isTabletOrMobile
            ? {width: "100%", height: "400px"}
            : {width: "800px", height: "800px"}
        }
        className={storePosition.map}
      ></div>
      <div className={storePosition.addressBlock}>
        <input ref={addressInputRef} className={storePosition.addressInput} />
        <button
          className={storePosition.addressButton}
          onClick={() => {
            changeAddress();
          }}
          data-testid="changeAddress"
        >
          주소 설정
        </button>
      </div>
      <div className={storePosition.submitButtonBlock}>
        <button
          className={storePosition.submitButton}
          onClick={() => {
            changePosition();
          }}
          data-testid="changePosition"
        >
          이 위치로 장소 수정
        </button>
      </div>
    </div>
  );
};

export default UpdateStorePosition;
