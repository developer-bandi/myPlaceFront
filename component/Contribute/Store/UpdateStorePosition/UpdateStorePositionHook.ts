import {useRouter} from "next/router";
import {useEffect, useRef, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {axiosUpdateStorePosition} from "../../../../lib/commonFn/api";
import {RootReducer} from "../../../../store";
import {setPosition} from "../../../../store/reducers/AddStorePosition/Reducer";
import {getStoreInfo} from "../../../../store/reducers/storeInfo/Reducer";
declare global {
  interface Window {
    kakao: any;
  }
}
const useUpdateStorePosition = () => {
  const mapref = useRef<HTMLDivElement>(null);
  const addressInputRef = useRef<HTMLInputElement>(null);
  const [makedmap, setMakedMap] = useState<any>();
  const [marker, setMarker] = useState<any>();
  const [loading, setLoading] = useState<boolean>();
  const [latitude, setLatitude] = useState("");
  const [longitude, setLongitude] = useState("");
  const [address, setAddress] = useState<string>();

  const dispatch = useDispatch();
  const router = useRouter();
  const {id} = router.query;
  const existPosition = useSelector(
    (state: RootReducer) => state.storeInfo.content?.storeInfo
  );

  useEffect(() => {
    if (existPosition !== undefined) {
      setLatitude(existPosition.latitude);
      setLongitude(existPosition.longitude);
      setAddress(existPosition.address);
      if (addressInputRef.current !== null) {
        addressInputRef.current.value = existPosition.address;
      }
    } else {
      alert("가게 정보가 없습니다. 가게를 선택해주세요");
      router.push("/findplace");
    }
  }, []);

  useEffect(() => {
    const mapScript = document.createElement("script");
    mapScript.type = "text/javascript";
    mapScript.async = true;
    mapScript.src = `//dapi.kakao.com/v2/maps/sdk.js?appkey=3efe55ecfaec2061baf17a762ef01372&autoload=false&libraries=services,clusterer`;
    document.head.appendChild(mapScript);
    const onLoadMap = () => {
      window.kakao.maps.load(() => {
        const options = {
          center: new window.kakao.maps.LatLng(
            35.23139384418825,
            129.086333411491
          ),
          level: 3,
        };
        const map = new window.kakao.maps.Map(mapref.current, options);
        const zoomControl = new window.kakao.maps.ZoomControl();
        const newMarker = new window.kakao.maps.Marker();
        map.addControl(zoomControl, window.kakao.maps.ControlPosition.TOPRIGHT);
        setMakedMap(map);
        setMarker(newMarker);
        setLoading(true);
      });
    };
    mapScript.addEventListener("load", onLoadMap);
    return () => mapScript.removeEventListener("load", onLoadMap);
  }, []);

  useEffect(() => {
    if (loading) {
      const clickHandeler = (mouseEvent: {
        latLng: {
          Ma: string;
          La: string;
          getLng: () => void;
          getLat: () => void;
        };
      }) => {
        const geocoder = new window.kakao.maps.services.Geocoder();

        const searchDetailAddrFromCoords = (
          coords: {getLng: () => void; getLat: () => void},
          callback: (
            result: {
              road_address: {address_name: string};
              address: {address_name: string};
            }[],
            status: string
          ) => void
        ) => {
          geocoder.coord2Address(coords.getLng(), coords.getLat(), callback);
        };
        searchDetailAddrFromCoords(
          mouseEvent.latLng,
          function (
            result: {
              road_address: {address_name: string};
              address: {address_name: string};
            }[],
            status: string
          ) {
            if (status === window.kakao.maps.services.Status.OK) {
              marker.setPosition(mouseEvent.latLng);
              marker.setMap(makedmap);

              if (addressInputRef.current !== null) {
                if (result[0].road_address) {
                  addressInputRef.current.value =
                    result[0].road_address.address_name;

                  setLatitude(mouseEvent.latLng.Ma);
                  setLongitude(mouseEvent.latLng.La);
                  setAddress(result[0].road_address.address_name);
                } else {
                  addressInputRef.current.value =
                    result[0].address.address_name;
                  setLatitude(mouseEvent.latLng.Ma);
                  setLongitude(mouseEvent.latLng.La);
                  setAddress(result[0].address.address_name);
                }
              }
            }
          }
        );
      };
      window.kakao.maps.event.addListener(makedmap, "click", clickHandeler);
      if (
        latitude !== undefined &&
        longitude !== undefined &&
        address !== undefined
      ) {
        const placePosition = new window.kakao.maps.LatLng(latitude, longitude);
        marker.setPosition(placePosition);
        marker.setMap(makedmap);
        makedmap.setCenter(placePosition);
      }
    }
  }, [loading]);

  const changeAddress = () => {
    if (addressInputRef.current !== null) {
      const geocoder = new window.kakao.maps.services.Geocoder();
      geocoder.addressSearch(
        addressInputRef.current.value,
        function (result: {x: string; y: string}[], status: string) {
          if (
            status === window.kakao.maps.services.Status.OK &&
            addressInputRef.current !== null
          ) {
            var coords = new window.kakao.maps.LatLng(result[0].y, result[0].x);
            marker.setPosition(coords);
            marker.setMap(makedmap);
            makedmap.setCenter(coords);
            dispatch(
              setPosition({
                latitude: result[0].y,
                longitude: result[0].x,
                address: addressInputRef.current.value,
              })
            );
          } else if (status === window.kakao.maps.services.Status.ZERO_RESULT) {
            alert("주소를 정확하게 입력해주세요");
            setLatitude("");
            setLongitude("");
            setAddress("");
            if (addressInputRef.current !== null) {
              addressInputRef.current.value = "";
            }
            marker.setMap(null);
          } else {
            alert("서버에 에러가 발생하였습니다");
          }
        }
      );
    }
  };

  const changePosition = async () => {
    console.log(typeof id, typeof latitude, typeof longitude, typeof address);
    if (
      typeof id === "string" &&
      latitude !== "" &&
      longitude !== "" &&
      typeof address === "string"
    ) {
      try {
        console.log("test111");
        await axiosUpdateStorePosition(id, latitude, longitude, address);
        console.log("test");
        dispatch(getStoreInfo(Number(id)));
        alert("정상적으로 수정되었습니다");
        router.push("/findplace");
      } catch (err) {
        alert("서버에 오류가 발생했습니다");
      }
    }
  };

  return {
    mapref,
    addressInputRef,
    changeAddress,
    changePosition,
    latitude,
    longitude,
    address,
    setMarker,
    setMakedMap,
    setAddress,
    setLatitude,
    setLongitude,
  };
};

export default useUpdateStorePosition;
