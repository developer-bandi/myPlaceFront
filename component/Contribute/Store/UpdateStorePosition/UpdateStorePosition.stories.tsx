import React, { useEffect, useRef } from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import UpdateStorePosition from "./UpdateStorePosition";
import { action } from "@storybook/addon-actions";
import { useIsTabletOrMobile } from "../../../../hooks/mediaQuery";

export default {
  title: "Contribute/Store/UpdateStorePosition",
  component: UpdateStorePosition,
} as ComponentMeta<typeof UpdateStorePosition>;

const Template: ComponentStory<typeof UpdateStorePosition> = (arg) => {
  const isTabletOrMobile = useIsTabletOrMobile();
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    setTimeout(() => {
      if (ref.current !== null) {
        const div = document.createElement("div");
        div.style.backgroundColor = "black";
        div.style.width = "100%";
        div.style.height = "100%";
        ref.current.appendChild(div);
      }
    }, 2000);
  });

  return (
    <UpdateStorePosition
      {...arg}
      isTabletOrMobile={isTabletOrMobile}
      mapref={ref}
    />
  );
};

export const generalType = Template.bind({});
generalType.args = {
  changeAddress: action("changeAddress"),
  changePosition: action("changePosition"),
};
