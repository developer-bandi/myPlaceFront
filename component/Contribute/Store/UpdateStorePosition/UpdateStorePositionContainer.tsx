import { useIsTabletOrMobile } from "../../../../hooks/mediaQuery";
import UpdateStorePosition from "./UpdateStorePosition";
import useUpdateStorePosition from "./UpdateStorePositionHook";

const UpdateStorePositionContainer = ({}) => {
  const { mapref, addressInputRef, changeAddress, changePosition } =
    useUpdateStorePosition();
  const isTabletOrMobile = useIsTabletOrMobile();

  return (
    <UpdateStorePosition
      mapref={mapref}
      addressInputRef={addressInputRef}
      changeAddress={changeAddress}
      changePosition={changePosition}
      isTabletOrMobile={isTabletOrMobile}
    />
  );
};

export default UpdateStorePositionContainer;
