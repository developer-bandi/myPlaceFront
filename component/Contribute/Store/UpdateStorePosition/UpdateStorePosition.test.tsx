import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
  waitFor,
} from "@testing-library/react";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import newReact from "react";
import * as React from "react";
import { useRouter } from "next/router";
import UpdateStorePosition from "./UpdateStorePosition";
import useUpdateStorePosition from "./UpdateStorePositionHook";
import axios from "axios";

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));
const mockRouter = {
  push: jest.fn(),
  query: { id: "1" },
};
(useRouter as jest.Mock).mockReturnValue(mockRouter);

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("UpdateStorePosition Hook 테스트", () => {
  let infoBlank = true;
  const mockStore = configureMockStore();
  const StorePositionMockStore = mockStore({
    storeInfo: {
      content: {
        storeInfo: {
          address: "testAddress",
          latitude: "testLatitude",
          longitude: "testLongitude",
        },
      },
    },
  });
  const SettedStorePositionMockStore = configureMockStore()({
    storeInfo: {
      content: {},
    },
  });
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <Provider
      store={infoBlank ? StorePositionMockStore : SettedStorePositionMockStore}
    >
      {children}
    </Provider>
  );

  describe("데이터 처음 받아오는 useEffect 테스트", () => {
    it("데이터가 있는 경우", async () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      await waitFor(() => {
        expect(result.current.latitude).toBe("testLatitude");
        expect(result.current.longitude).toBe("testLongitude");
        expect(result.current.address).toBe("testAddress");
        expect(result.current.addressInputRef.current).toEqual({
          value: "testAddress",
        });
      });
    });
    it("데이터가 없는 경우", async () => {
      infoBlank = false;
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      window.alert = jest.fn();
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      expect(window.alert).toBeCalledWith(
        "가게 정보가 없습니다. 가게를 선택해주세요"
      );
      expect(mockRouter.push).toBeCalledWith("/findplace");
    });
  });

  describe("changeAddress 테스트", () => {
    it("정보를 제대로 찾은 경우", () => {
      infoBlank = true;
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{ x: "testx", y: "testy" }], "ok");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      const setPositionMock = jest.fn();
      const setMapMock = jest.fn();
      const setCenterMock = jest.fn();
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      act(() => {
        result.current.setMarker({
          setPosition: setPositionMock,
          setMap: setMapMock,
        });
      });
      act(() => {
        result.current.setMakedMap({
          setCenter: setCenterMock,
        });
      });
      act(() => {
        result.current.changeAddress();
      });
      expect(StorePositionMockStore.getActions()[0]).toStrictEqual({
        type: "addStorePosition/setPosition",
        payload: {
          address: "testAddress",
          latitude: "testy",
          longitude: "testx",
        },
      });
    });
    it("결과가 없는 경우", () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{ x: "testx", y: "testy" }], "zero_result");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      window.alert = jest.fn();
      const setPositionMock = jest.fn();
      const setMapMock = jest.fn();
      const setCenterMock = jest.fn();
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      act(() => {
        result.current.setMarker({
          setPosition: setPositionMock,
          setMap: setMapMock,
        });
      });
      act(() => {
        result.current.setMakedMap({
          setCenter: setCenterMock,
        });
      });
      act(() => {
        result.current.changeAddress();
      });
      expect(window.alert).toBeCalledWith("주소를 정확하게 입력해주세요");
      expect(result.current.address).toBe("");
      expect(result.current.latitude).toBe("");
      expect(result.current.longitude).toBe("");
      expect(result.current.addressInputRef.current).toEqual({ value: "" });
      expect(setMapMock.mock.calls[0][0]).toBe(null);
    });
    it("에러 발생", () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      const kakao = {
        maps: {
          services: {
            Geocoder: jest.fn(() => {
              return {
                addressSearch: (input: string, callback: Function) => {
                  callback([{ x: "testx", y: "testy" }], "error");
                },
              };
            }),
            Status: {
              OK: "ok",
              ZERO_RESULT: "zero_result",
              ERROR: "error",
            },
          },
          LatLng: jest.fn(() => "test"),
        },
      };
      window.kakao = kakao as any;
      window.alert = jest.fn();
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      act(() => {
        result.current.changeAddress();
      });
      expect(window.alert).toBeCalledWith("서버에 에러가 발생하였습니다");
    });
  });

  describe("changePosition 함수 테스트", () => {
    it("값이 채워져 있는 경우", async () => {
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      mockedAxios.patch.mockImplementation(() =>
        Promise.resolve({ status: 200, data: "test" })
      );
      window.alert = jest.fn();
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      act(() => {
        result.current.setAddress("test");
        result.current.setLatitude("test");
        result.current.setLongitude("test");
      });
      await act(async () => {
        await result.current.changePosition();
      });
      expect(window.alert).toBeCalledWith("정상적으로 수정되었습니다");
      expect(StorePositionMockStore.getActions()[1]).toEqual({
        payload: 1,
        type: "storeInfo/getStoreInfo",
      });
      expect(mockRouter.push).toBeCalledWith("/findplace");
    });

    it("값이 비어 있는 경우", async () => {
      infoBlank = true;
      window.alert = jest.fn();
      const useRefSpy = jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({ current: { value: "test" } });
      window.alert = jest.fn();
      mockedAxios.patch.mockImplementation(() =>
        Promise.reject({ status: 500, data: "error" })
      );
      const { result } = renderHook(() => useUpdateStorePosition(), {
        wrapper,
      });
      await act(async () => {
        await result.current.changePosition();
      });
      expect(window.alert).toBeCalledWith("서버에 오류가 발생했습니다");
    });
  });
});
