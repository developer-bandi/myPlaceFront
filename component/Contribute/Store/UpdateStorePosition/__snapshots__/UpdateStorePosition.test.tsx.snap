// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <div
    class="mainBlock"
  >
    <h1
      class="title"
    >
      장소 수정
    </h1>
    <div
      class="map"
      style="width: 800px; height: 800px;"
    />
    <div
      class="addressBlock"
    >
      <input
        class="addressInput"
      />
      <button
        class="addressButton"
        data-testid="changeAddress"
      >
        주소 설정
      </button>
    </div>
    <div
      class="submitButtonBlock"
    >
      <button
        class="submitButton"
        data-testid="changePosition"
      >
        이 위치로 장소 수정
      </button>
    </div>
  </div>
</div>
`;
